File: runtime/symdim.py
# Ethan Doughty
# symdim.py
"""Symbolic dimension type for MATLAB static analysis.

Defines SymDim — a canonical multivariate polynomial representation
for symbolic matrix dimensions.
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional
from fractions import Fraction

# A monomial: sorted tuple of (variable_name, exponent) pairs.
# Empty tuple = constant monomial (1).
Monomial = tuple  # tuple[tuple[str, int], ...]


@dataclass(frozen=True)
class SymDim:
    """Canonical symbolic dimension expression (multivariate polynomial).

    Sorted tuple of (monomial, coefficient) pairs.
    Canonical by construction: sorted, collected, no zero coefficients.
    """
    _terms: tuple  # tuple[tuple[Monomial, Fraction], ...]

    @staticmethod
    def const(value: int) -> "SymDim":
        """Create a constant symbolic dimension."""
        if value == 0:
            return SymDim(_terms=())
        return SymDim(_terms=(((), Fraction(value)),))

    @staticmethod
    def var(name: str) -> "SymDim":
        """Create a variable symbolic dimension."""
        return SymDim(_terms=(((((name, 1),), Fraction(1))),))

    @staticmethod
    def zero() -> "SymDim":
        """Create zero symbolic dimension."""
        return SymDim(_terms=())

    def __add__(self, other: "SymDim") -> "SymDim":
        """Add two symbolic dimensions."""
        # Merge terms from both polynomials
        term_dict = {}
        for mono, coeff in self._terms:
            term_dict[mono] = term_dict.get(mono, 0) + coeff
        for mono, coeff in other._terms:
            term_dict[mono] = term_dict.get(mono, 0) + coeff

        # Remove zero coefficients and sort
        terms = [(mono, coeff) for mono, coeff in term_dict.items() if coeff != 0]
        terms.sort(key=lambda t: self._mono_key(t[0]))
        return SymDim(_terms=tuple(terms))

    def __neg__(self) -> "SymDim":
        """Negate a symbolic dimension."""
        return SymDim(_terms=tuple((mono, -coeff) for mono, coeff in self._terms))

    def __sub__(self, other: "SymDim") -> "SymDim":
        """Subtract two symbolic dimensions."""
        return self + (-other)

    def __truediv__(self, other: "SymDim") -> "SymDim":
        """Divide two symbolic dimensions (returns rational coefficients).

        Only supports division by constants.
        """
        # Get constant divisor
        divisor = other.const_value()
        if divisor is None:
            raise ValueError("Division only supported for constant divisors")
        if divisor == 0:
            raise ValueError("Division by zero")

        # Scale all coefficients by 1/divisor
        frac_divisor = Fraction(divisor)
        new_terms = tuple((mono, coeff / frac_divisor) for mono, coeff in self._terms)
        return SymDim(_terms=new_terms)

    def __mul__(self, other: "SymDim") -> "SymDim":
        """Multiply two symbolic dimensions."""
        # Cross-product of terms
        term_dict = {}
        for mono1, coeff1 in self._terms:
            for mono2, coeff2 in other._terms:
                # Multiply monomials: combine variable exponents
                var_dict = {}
                for var, exp in mono1:
                    var_dict[var] = var_dict.get(var, 0) + exp
                for var, exp in mono2:
                    var_dict[var] = var_dict.get(var, 0) + exp
                # Sort variables to get canonical monomial
                new_mono = tuple(sorted(var_dict.items()))
                # Add to term dict
                term_dict[new_mono] = term_dict.get(new_mono, 0) + coeff1 * coeff2

        # Remove zeros and sort
        terms = [(mono, coeff) for mono, coeff in term_dict.items() if coeff != 0]
        terms.sort(key=lambda t: self._mono_key(t[0]))
        return SymDim(_terms=tuple(terms))

    def is_const(self) -> bool:
        """Check if this is a constant (no variables)."""
        return len(self._terms) == 0 or (len(self._terms) == 1 and self._terms[0][0] == ())

    def const_value(self) -> Optional[int]:
        """Return constant value if this is a constant, else None."""
        if len(self._terms) == 0:
            return 0
        if len(self._terms) == 1 and self._terms[0][0] == ():
            coeff = self._terms[0][1]
            # Only return int if denominator is 1
            if isinstance(coeff, Fraction) and coeff.denominator == 1:
                return int(coeff)
        return None

    @staticmethod
    def _mono_key(mono: Monomial) -> tuple:
        """Sort key for monomials: higher degree first, then alphabetical."""
        degree = sum(exp for var, exp in mono)
        return (-degree, mono)

    def variables(self) -> set:
        """Return set of variable names in this symbolic dimension.

        Returns:
            Set of variable name strings appearing in any term.
        """
        var_set = set()
        for mono, _ in self._terms:
            for var_name, _ in mono:
                var_set.add(var_name)
        return var_set

    def substitute(self, bindings: dict) -> "SymDim":
        """Replace bound variables with concrete values; keep free vars symbolic.

        Args:
            bindings: {var_name: int} mapping variables to concrete values.

        Returns:
            New SymDim with bound vars replaced. Free vars remain symbolic.
        """
        result = SymDim.zero()
        for mono, coeff in self._terms:
            # Start with the coefficient as a constant SymDim
            term_value = SymDim.const(1) if coeff == 1 else SymDim(_terms=(((), coeff),))
            for var_name, exp in mono:
                if var_name in bindings:
                    # Substitute concrete value: multiply term by bindings[var]^exp
                    concrete = Fraction(bindings[var_name]) ** exp
                    term_value = SymDim(_terms=tuple(
                        (m, c * concrete) for m, c in term_value._terms
                    ))
                else:
                    # Keep free variable symbolic: var_factor = SymDim.var(var_name)
                    var_factor = SymDim.var(var_name)
                    # Raise to exp by repeated multiplication
                    factor = SymDim.const(1)
                    for _ in range(exp):
                        factor = factor * var_factor
                    term_value = term_value * factor
            result = result + term_value
        return result

    def evaluate(self, bindings: dict) -> Optional[int]:
        """Evaluate to int if all variables are bound after substitution.

        Args:
            bindings: {var_name: int} mapping variables to concrete values.

        Returns:
            Integer value if fully ground, None if free variables remain or
            result is non-integer (e.g., n/2 with n=3 gives 3/2).
        """
        result = self.substitute(bindings)
        return result.const_value()

    def __str__(self) -> str:
        """Format polynomial with degree-descending, alphabetical display."""
        if len(self._terms) == 0:
            return "0"

        parts = []
        for mono, coeff in self._terms:
            # Format coefficient (handle Fraction display)
            if isinstance(coeff, Fraction):
                if coeff.denominator == 1:
                    coeff_str = str(int(coeff))
                    coeff_int = int(coeff)
                else:
                    # Rational coefficient
                    coeff_str = f"({coeff.numerator}/{coeff.denominator})"
                    coeff_int = None
            else:
                coeff_str = str(coeff)
                coeff_int = coeff

            if mono == ():
                # Constant term
                parts.append(coeff_str)
            else:
                # Variable term
                var_parts = []
                for var, exp in mono:
                    if exp == 1:
                        var_parts.append(var)
                    else:
                        var_parts.append(f"{var}^{exp}")
                var_str = "*".join(var_parts)

                # Special case for single-var single-coeff rational: n/2 instead of (1/2)*n
                if (isinstance(coeff, Fraction) and coeff.denominator != 1 and
                    len(mono) == 1 and mono[0][1] == 1 and coeff.numerator == 1):
                    # Format as "n/d" instead of "(1/d)*n"
                    parts.append(f"{var_str}/{coeff.denominator}")
                elif coeff_int == 1:
                    parts.append(var_str)
                elif coeff_int == -1:
                    parts.append(f"-{var_str}")
                else:
                    parts.append(f"{coeff_str}*{var_str}")

        # Join with + or - (handle negative coefficients)
        result = parts[0]
        for part in parts[1:]:
            if part.startswith("-"):
                result += part  # Already has minus sign
            else:
                result += "+" + part

        return result

File: fsharp/SymDim.fs
module SymDim

open System.Collections.Generic

// ---------------------------------------------------------------------------
// Rational number (numerator/denominator as int64, always normalized)
// ---------------------------------------------------------------------------

[<Struct; CustomEquality; NoComparison>]
type Rational private (num: int64, den: int64) =
    member _.Numerator   = num
    member _.Denominator = den

    static member private Normalize(n: int64, d: int64) =
        if d = 0L then failwith "Rational: division by zero"
        let sign = if d < 0L then -1L else 1L
        let n' = n * sign
        let d' = d * sign
        let rec gcd a b = if b = 0L then a else gcd b (a % b)
        let g = gcd (abs n') d'
        Rational(n' / g, d' / g)

    static member Create(n: int64, d: int64) = Rational.Normalize(n, d)
    static member FromInt(n: int) = Rational.Normalize(int64 n, 1L)
    static member Zero = Rational.Normalize(0L, 1L)
    static member One  = Rational.Normalize(1L, 1L)

    static member (+)(a: Rational, b: Rational) =
        Rational.Normalize(a.Numerator * b.Denominator + b.Numerator * a.Denominator,
                           a.Denominator * b.Denominator)
    static member (-)(a: Rational, b: Rational) =
        Rational.Normalize(a.Numerator * b.Denominator - b.Numerator * a.Denominator,
                           a.Denominator * b.Denominator)
    static member (*)(a: Rational, b: Rational) =
        Rational.Normalize(a.Numerator * b.Numerator, a.Denominator * b.Denominator)
    static member (/)(a: Rational, b: Rational) =
        if b.Numerator = 0L then failwith "Rational: division by zero"
        Rational.Normalize(a.Numerator * b.Denominator, a.Denominator * b.Numerator)
    static member (~-)(a: Rational) = Rational.Normalize(-a.Numerator, a.Denominator)

    member this.IsZero    = this.Numerator = 0L
    member this.IsInteger = this.Denominator = 1L
    member this.ToInt64   = this.Numerator  // only valid when IsInteger

    override this.Equals(obj) =
        match obj with
        | :? Rational as r -> this.Numerator = r.Numerator && this.Denominator = r.Denominator
        | _ -> false
    override this.GetHashCode() = hash (this.Numerator, this.Denominator)

// ---------------------------------------------------------------------------
// Monomial: sorted list of (variable_name, exponent) pairs
// Empty list = constant monomial (1)
// ---------------------------------------------------------------------------

type Monomial = (string * int) list

// Sort key for monomials: higher degree first, then lexicographic
let private monoKey (mono: Monomial) : int * Monomial =
    let degree = mono |> List.sumBy snd
    (-degree, mono)

// ---------------------------------------------------------------------------
// SymDim: multivariate polynomial over rational coefficients
// Represented as a sorted association list: (Monomial * Rational) list
// Canonical: sorted by monoKey, no zero coefficients
// ---------------------------------------------------------------------------

type SymDim = { _terms: (Monomial * Rational) list }

// ---------------------------------------------------------------------------
// Module SymDim: constructors and operations
// ---------------------------------------------------------------------------

module SymDim =

    let zero () : SymDim = { _terms = [] }

    let const' (value: int) : SymDim =
        if value = 0 then zero ()
        else { _terms = [ ([], Rational.FromInt value) ] }

    let var (name: string) : SymDim =
        { _terms = [ ([name, 1], Rational.One) ] }

    // Collect terms from a dictionary into canonical sorted list
    let private canonicalize (termDict: Dictionary<Monomial, Rational>) : (Monomial * Rational) list =
        termDict
        |> Seq.choose (fun kv -> if kv.Value.IsZero then None else Some (kv.Key, kv.Value))
        |> Seq.toList
        |> List.sortBy (fun (mono, _) -> monoKey mono)

    let add (a: SymDim) (b: SymDim) : SymDim =
        let d = Dictionary<Monomial, Rational>(HashIdentity.Structural)
        for (mono, coeff) in a._terms do
            if d.ContainsKey(mono) then d.[mono] <- d.[mono] + coeff
            else d.[mono] <- coeff
        for (mono, coeff) in b._terms do
            if d.ContainsKey(mono) then d.[mono] <- d.[mono] + coeff
            else d.[mono] <- coeff
        { _terms = canonicalize d }

    let neg (a: SymDim) : SymDim =
        { _terms = a._terms |> List.map (fun (mono, coeff) -> (mono, -coeff)) }

    let sub (a: SymDim) (b: SymDim) : SymDim = add a (neg b)

    let mul (a: SymDim) (b: SymDim) : SymDim =
        let d = Dictionary<Monomial, Rational>(HashIdentity.Structural)
        for (mono1, coeff1) in a._terms do
            for (mono2, coeff2) in b._terms do
                let varDict = Dictionary<string, int>()
                for (v, e) in mono1 do
                    if varDict.ContainsKey(v) then varDict.[v] <- varDict.[v] + e
                    else varDict.[v] <- e
                for (v, e) in mono2 do
                    if varDict.ContainsKey(v) then varDict.[v] <- varDict.[v] + e
                    else varDict.[v] <- e
                let newMono =
                    varDict |> Seq.map (fun kv -> (kv.Key, kv.Value)) |> Seq.toList |> List.sortBy fst
                let newCoeff = coeff1 * coeff2
                if d.ContainsKey(newMono) then d.[newMono] <- d.[newMono] + newCoeff
                else d.[newMono] <- newCoeff
        { _terms = canonicalize d }

    let div (a: SymDim) (divisor: int) : SymDim =
        if divisor = 0 then failwith "SymDim.div: division by zero"
        let fracDiv = Rational.FromInt divisor
        { _terms = a._terms |> List.map (fun (mono, coeff) -> (mono, coeff / fracDiv)) }

    let divBySymDim (a: SymDim) (b: SymDim) : SymDim =
        match b._terms with
        | [([], c)] when c.IsInteger -> div a (int c.ToInt64)
        | _ -> failwith "SymDim: division only supported for constant divisors"

    let isConst (s: SymDim) : bool =
        s._terms.IsEmpty || (s._terms.Length = 1 && fst s._terms.[0] = [])

    let constValue (s: SymDim) : int option =
        match s._terms with
        | [] -> Some 0
        | [([], c)] when c.IsInteger -> Some (int c.ToInt64)
        | _ -> None

    let variables (s: SymDim) : Set<string> =
        s._terms
        |> List.collect (fun (mono, _) -> mono |> List.map fst)
        |> Set.ofList

    let substitute (bindings: Map<string, int>) (s: SymDim) : SymDim =
        let mutable result = zero ()
        for (mono, coeff) in s._terms do
            let mutable termValue =
                if coeff = Rational.One then const' 1
                else { _terms = [([], coeff)] }
            for (varName, exp) in mono do
                match Map.tryFind varName bindings with
                | Some concreteVal ->
                    let concrete = Rational.Create(int64 (pown concreteVal exp), 1L)
                    termValue <- { _terms = termValue._terms |> List.map (fun (m, c) -> (m, c * concrete)) }
                | None ->
                    let varFactor = var varName
                    let mutable factor = const' 1
                    for _ in 1..exp do
                        factor <- mul factor varFactor
                    termValue <- mul termValue factor
            result <- add result termValue
        result

    let evaluate (bindings: Map<string, int>) (s: SymDim) : int option =
        constValue (substitute bindings s)

    let private formatMono (mono: Monomial) : string =
        mono
        |> List.map (fun (v, e) -> if e = 1 then v else v + "^" + string e)
        |> String.concat "*"

    let toString (s: SymDim) : string =
        if s._terms.IsEmpty then "0"
        else
            let parts =
                s._terms |> List.map (fun (mono, coeff) ->
                    if mono = [] then
                        // Constant term
                        if coeff.IsInteger then string coeff.Numerator
                        else "(" + string coeff.Numerator + "/" + string coeff.Denominator + ")"
                    else
                        let varStr = formatMono mono
                        // Special case: single-var rational n/d where numerator=1 -> "n/d"
                        if not coeff.IsInteger && mono.Length = 1 && snd mono.[0] = 1 && coeff.Numerator = 1L then
                            varStr + "/" + string coeff.Denominator
                        elif coeff = Rational.One then
                            varStr
                        elif coeff = -Rational.One then
                            "-" + varStr
                        elif coeff.IsInteger then
                            string coeff.Numerator + "*" + varStr
                        else
                            "(" + string coeff.Numerator + "/" + string coeff.Denominator + ")*" + varStr
                )
            let mutable result = parts.[0]
            for part in parts |> List.tail do
                if part.StartsWith("-") then result <- result + part
                else result <- result + "+" + part
            result
